### ISSUE TITLE
lookup href, con contatenation, empty values

### ISSUE THREAD

**Issue by petervanderstok** [core-wg/resource-directory#34]
2.  I am unclear if the context is what should be returned as the link
address for an endpoint or if it is just the scheme + authority (i.e.
omit
any path portion).

I find the concatenation of "con" and "href" values, as currently
suggested in
the examples, also difficult to understand.
IMO the href value should be returned between <> brackets, and the con as
an additional attribute.

I would find that logical, even if it makes the client side more difficult
to build the entire URI.

3.  For the domain and group lookups, is the href an empty string or
is it absent?  This does not make a different for the standard link
formation, but it does make a difference for the cbor and json
formats.

@ Michael can you answer, being involved in the link-json document

4.  I assume that the 'et' parameter should be returned on all end
points.
et can have no value; IMO do not return anything.

Do you mean, do not synthesis a value from the registration and add it?
That would be fine with me.

5.  How would an empty domain name be returned (this is my current
pre-configured default domain) - as an empty string or as just the
attribute name - i.e  d  or d="" ?

Actually the d= can have no value, and IMO like et, do not return anything
when there is no value.

**Comment by mjkoster** [comment]
1. What's the question?

2. Let's be careful not to confuse endpoint lookup with resource lookup. The question is about endpoint lookup. (also note that the issue contains questions and answers copied from an email)
For endpoint lookups, if we allow "con" to contain a base path for all registered resources, we should probably include the base path in the endpoint as returned by endpoint lookup. The rationale for this is that a device could register more than one unique endpoint (using different base paths) and manage them separately in the RD. If this is unclear, I can make some examples.

3. Domain lookups should be removed, as domain is an attribute of an endpoint and there is no domain list. 
Group lookups should return the group resource location in href (see my comment in issue #35) but please note I think the group lookup should return the  multicast address in "con", in the same way as it was registered rather than in "href".

4. "et" is optional and does not need to be returned if there was no registered value. 

5. "domain" is optional and is now defined as an attribute of a registration. No fake domain is needed if there was none supplied on registration

**Comment by chrysn** [comment]
ad 2: As I understand the path component of an endpoint address, this is something we allowed back when CoAP-over-Websockets needed a path on the web server, and left in to not arbitrarily limit the URI syntax -- so before Michael's comment, I'd have assumed that anything with a non-empty path would probably come from an exotic (future) URI scheme and probably not a good idea unless well thought-out. Michale, I'd be curious to see an illustrative example of this.

ad 3.: agree with Michael on Domain (and need to revisit the current state on Group); at any rate, these should not be empty `<>`s (see https://github.com/core-wg/resource-directory/issues/21).

ad 4 and 5: IMO both should be shown in the endpoint lookup if given explicitly, and not given at all when none is given or configured (ie. not `<...>;d` or `<...>;d=` but just `<...>`).

**Comment by chrysn** [comment]
Crossreference: This issue stems from the thread archived at <https://mailarchive.ietf.org/arch/msg/core/G4Kx2YFQ8xVPCOWR4dfVACMgwGo> which has diverged a bit from the discussion here; trying to tie them together again...

**Comment by chrysn** [comment]
Hello Jim and Peter,

On Fri, Sep 15, 2017 at 04:14:51PM -0700, Jim Schaad wrote:
> > > 2.  I am unclear if the context is what should be returned as the
> > > link address for an endpoint or if it is just the scheme +
> > > authority (i.e.  omit any path portion).
> > 
> > I find the concatenation of "con" and "href" values, as currently
> > suggested in the examples, also difficult to understand.  IMO the
> > href value should be returned between <> brackets, and the con as an
> > additional attribute.
> 
> I would find that logical, even if it makes the client side more difficult
> to build the entire URI.

as I view the resource lookup, that interface should be usable for
clients without awareness of what a resource directory is. (For example,
a temperature display that is used to fetch all temperatures from a
given device should be able to use
`coap://configued-rd/lkp/res?d=my-domain` as its configured address and,
by the same rules by which it'd discover a single device's temperatures,
display all the domains').

For that, and to faithfully represent not only the links but also the
link relations, a lookup result should be

    </temp/1>;rt="temp.inside";anchor="coap://first-device",
    </temp/1>;rt="temp.inside";anchor="coap://second-device"

Thus, href stays as it was registered, all attributes stay as they are
(including the implied rel="hosts"), but the implicit
anchor="coap://first-device";rel="hosts" that all lines from a lookup to
coap://first-device/.well-known/core would carry is made explicit where
different defaults would take over otherwise.

In other words, the con= value of the registered endpoint is expressed
as anchor in the EP lookup because that precisely fits its meaning.

(AFAIR this is something we've also talked about in Prague, and I should
probably write a pull request that brings the examples in line).

Best regards
Christian

-- 
There's always a bigger fish.
  -- Qui-Gon Jinn

**Comment by mjkoster** [comment]
Yes, Christian provides a good summary. I think if we can work through some examples it will be clear.

**Comment by chrysn** [comment]
I think that all issues brought up here are addressed either by the now consistently explained reference resolution rules, or through the updated registration parameter representation and lookup.
